// @graphql-codegen/typescript-operations/index.esm.js
import { isEnumType, isNonNullType, isListType, concatAST, Kind, visit } from 'graphql';
import { BaseSelectionSetProcessor, BaseDocumentsVisitor, getConfigValue, PreResolveTypesProcessor, SelectionSetToObject, optimizeOperations } from '@graphql-codegen/visitor-plugin-common';
import { TypeScriptOperationVariablesToObject as TypeScriptOperationVariablesToObject$1 } from '@graphql-codegen/typescript';
import autoBind from 'auto-bind';

class TypeScriptOperationVariablesToObject extends TypeScriptOperationVariablesToObject$1 {
    formatTypeString(fieldType, isNonNullType, hasDefaultValue) {
        return fieldType;
    }
}

class TypeScriptSelectionSetProcessor extends BaseSelectionSetProcessor {
    transformPrimitiveFields(schemaType, fields) {
        if (fields.length === 0) {
            return [];
        }
        const parentName = (this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : '') +
            this.config.convertName(schemaType.name, {
                useTypesPrefix: true,
            });
        return [`Pick<${parentName}, ${fields.map(field => `'${field}'`).join(' | ')}>`];
    }
    transformTypenameField(type, name) {
        return [`{ ${name}: ${type} }`];
    }
    transformAliasesPrimitiveFields(schemaType, fields) {
        if (fields.length === 0) {
            return [];
        }
        const parentName = (this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : '') +
            this.config.convertName(schemaType.name, {
                useTypesPrefix: true,
            });
        return [
            `{ ${fields
                .map(aliasedField => {
                const value = aliasedField.fieldName === '__typename' ? `'${schemaType.name}'` : `${parentName}['${aliasedField.fieldName}']`;
                return `${this.config.formatNamedField(aliasedField.alias)}: ${value}`;
            })
                .join(', ')} }`,
        ];
    }
    transformLinkFields(fields) {
        if (fields.length === 0) {
            return [];
        }
        return [`{ ${fields.map(field => `${this.config.formatNamedField(field.alias || field.name)}: ${field.selectionSet}`).join(', ')} }`];
    }
}

class TypeScriptDocumentsVisitor extends BaseDocumentsVisitor {
    constructor(schema, config, allFragments) {
        super(config, {
            noExport: getConfigValue(config.noExport, false),
            avoidOptionals: typeof config.avoidOptionals === 'boolean' ? getConfigValue(config.avoidOptionals, false) : false,
            immutableTypes: getConfigValue(config.immutableTypes, false),
            nonOptionalTypename: getConfigValue(config.nonOptionalTypename, false),
        }, schema);
        autoBind(this);
        const clearOptional = (str) => {
            const prefix = this.config.namespacedImportName ? `${this.config.namespacedImportName}\.` : '';
            const rgx = new RegExp(`^${prefix}Maybe<(.*?)>$`, 'is');
            if (str.startsWith(`${this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : ''}Maybe`)) {
                return str.replace(rgx, '$1');
            }
            return str;
        };
        const wrapTypeWithModifiers = (baseType, type) => {
            const prefix = this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : '';
            if (isNonNullType(type)) {
                return clearOptional(wrapTypeWithModifiers(baseType, type.ofType));
            }
            else if (isListType(type)) {
                const innerType = wrapTypeWithModifiers(baseType, type.ofType);
                return `${prefix}Maybe<${this.config.immutableTypes ? 'ReadonlyArray' : 'Array'}<${innerType}>>`;
            }
            else {
                return `${prefix}Maybe<${baseType}>`;
            }
        };
        const processorConfig = {
            namespacedImportName: this.config.namespacedImportName,
            convertName: this.convertName.bind(this),
            enumPrefix: this.config.enumPrefix,
            scalars: this.scalars,
            formatNamedField: (name) => (this.config.immutableTypes ? `readonly ${name}` : name),
            wrapTypeWithModifiers,
        };
        const processor = new (config.preResolveTypes ? PreResolveTypesProcessor : TypeScriptSelectionSetProcessor)(processorConfig);
        this.setSelectionSetHandler(new SelectionSetToObject(processor, this.scalars, this.schema, this.convertName.bind(this), allFragments, this.config));
        const enumsNames = Object.keys(schema.getTypeMap()).filter(typeName => isEnumType(schema.getType(typeName)));
        this.setVariablesTransformer(new TypeScriptOperationVariablesToObject(this.scalars, this.convertName.bind(this), this.config.avoidOptionals, this.config.immutableTypes, this.config.namespacedImportName, enumsNames, this.config.enumPrefix, this.config.enumValues));
        this._declarationBlockConfig = {
            ignoreExport: this.config.noExport,
        };
    }
    getPunctuation(declarationKind) {
        return ';';
    }
}

const plugin = (schema, rawDocuments, config) => {
    const documents = config.flattenGeneratedTypes ? optimizeOperations(schema, rawDocuments) : rawDocuments;
    const allAst = concatAST(documents.map(v => v.document));
    const allFragments = [
        ...allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION).map(fragmentDef => ({ node: fragmentDef, name: fragmentDef.name.value, onType: fragmentDef.typeCondition.name.value, isExternal: false })),
        ...(config.externalFragments || []),
    ];
    const visitorResult = visit(allAst, {
        leave: new TypeScriptDocumentsVisitor(schema, config, allFragments),
    });
    const result = visitorResult.definitions.join('\n');
    if (config.globalNamespace) {
        return `
    declare global { 
      ${result} 
    }
          `;
    }
    else {
        return result;
    }
};

export { TypeScriptDocumentsVisitor, plugin };
