// @graphql-toolkit/git-loader/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const common = require('@graphql-toolkit/common');
const simplegit = _interopDefault(require('simple-git/promise'));
const graphqlTagPluck = require('@graphql-toolkit/graphql-tag-pluck');

// git:branch:path/to/file
function extractData(pointer) {
    const parts = pointer.replace(/^git\:/i, '').split(':');
    if (!parts || parts.length !== 2) {
        throw new Error('Schema pointer should match "git:branchName:path/to/file"');
    }
    return {
        ref: parts[0],
        path: parts[1],
    };
}
class GitLoader {
    loaderId() {
        return 'git-loader';
    }
    async canLoad(pointer) {
        return typeof pointer === 'string' && pointer.toLowerCase().startsWith('git:');
    }
    async load(pointer, options) {
        const { ref, path } = extractData(pointer);
        const git = simplegit();
        let content;
        try {
            content = await git.show([`${ref}:${path}`]);
        }
        catch (error) {
            throw new Error('Unable to load schema from git: ' + error);
        }
        if (/\.(gql|graphql)s?$/i.test(path)) {
            return common.parseGraphQLSDL(pointer, content, options);
        }
        if (/\.json$/i.test(path)) {
            return common.parseGraphQLJSON(pointer, content, options);
        }
        const rawSDL = await graphqlTagPluck.gqlPluckFromCodeString(pointer, content, options.pluckConfig);
        if (rawSDL) {
            return {
                location: pointer,
                rawSDL,
            };
        }
        throw new Error(`Invalid file extension: ${path}`);
    }
}

exports.GitLoader = GitLoader;
